File: tools/orchestrator/src/schemas/rankedSources.ts
import { z } from "zod";

export const RankedBulletSchema = z.object({
  text: z.string().min(1),
  evidence: z.array(z.string().min(1)).min(1),
  relevancy: z.number(),
  freshness: z.number(),
  score: z.number()
});

export type RankedBullet = z.infer<typeof RankedBulletSchema>;

export const RankedEntrySchema = z.object({
  id: z.string().min(1),
  kind: z.enum(["experience", "education", "project"]),
  group: z.string().min(1),
  title: z.string().min(1),
  role: z.string().optional(),
  location: z.string().optional(),
  duration: z.string().optional(),
  diploma: z.string().optional(),
  gpa: z.string().optional(),
  relevancy: z.number(),
  freshness: z.number(),
  score: z.number(),
  evidence: z.array(z.string().min(1)).default([]),
  bullets: z.array(RankedBulletSchema).default([])
});

export type RankedEntry = z.infer<typeof RankedEntrySchema>;

export const RankedSourcesSchema = z.object({
  schemaVersion: z.literal(1),
  generatedAt: z.string().min(1),
  job: z.object({
    path: z.string().min(1),
    company: z.string().optional(),
    focus: z.string().optional()
  }),
  weights: z.object({
    relevancy: z.number(),
    freshness: z.number()
  }),
  experience: z.array(RankedEntrySchema).default([]),
  education: z.array(RankedEntrySchema).default([]),
  projects: z.array(RankedEntrySchema).default([])
});

export type RankedSources = z.infer<typeof RankedSourcesSchema>;

File: backend/src/resume_orchestrator/schemas/ranked_sources.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Kind = Literal["experience", "education", "project"]


class RankedBullet(BaseModel):
    model_config = ConfigDict(extra="ignore")

    text: str = Field(min_length=1)
    evidence: list[str] = Field(min_length=1)
    relevancy: float
    freshness: float
    score: float


class RankedEntry(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str = Field(min_length=1)
    kind: Kind
    group: str = Field(min_length=1)
    title: str = Field(min_length=1)
    role: str | None = None
    location: str | None = None
    duration: str | None = None
    diploma: str | None = None
    gpa: str | None = None
    relevancy: float
    freshness: float
    score: float
    evidence: list[str] = Field(default_factory=list)
    bullets: list[RankedBullet] = Field(default_factory=list)


class RankedJob(BaseModel):
    model_config = ConfigDict(extra="ignore")

    path: str = Field(min_length=1)
    company: str | None = None
    focus: str | None = None


class RankedWeights(BaseModel):
    model_config = ConfigDict(extra="ignore")

    relevancy: float
    freshness: float


class RankedSources(BaseModel):
    model_config = ConfigDict(extra="ignore")

    schemaVersion: Literal[1] = 1
    generatedAt: str = Field(min_length=1)
    job: RankedJob
    weights: RankedWeights
    experience: list[RankedEntry] = Field(default_factory=list)
    education: list[RankedEntry] = Field(default_factory=list)
    projects: list[RankedEntry] = Field(default_factory=list)
